Add initial support for stm32

thom24 opened the pull request:
This pull request adds an initial support for stm32 boards.

For now only the stm32mp1 subtarget was added.
The first defined device is STM32MP135F-DK.

thom24 committed:
mac80211: enable CONFIG_BRCMFMAC_SDIO for stm32

All stm32 devices use BCM43430 Wi-Fi interface in SDIO mode.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
build: add generic build template for optee-os packages

Add a generic template include/optee-os.mk to build OP-TEE.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
optee-os-stm32: add OP-TEE for stm32 boards

Add OP-TEE for stm32 boards, with the first being STM32MP135F-DK.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
uboot-stm32: add U-Boot for stm32 boards

Add U-Boot for stm32 boards, with the first being STM32MP135F-DK.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
arm-trusted-firmware-stm32: add ATF for stm32 boards

Add ATF for stm32 boards, with the first being STM32MP135F-DK.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
stm32: add new stm32 target

New stm32 target introduces support for stm32mp1 based devices.
For now it includes an initial support of the STM32MP135F-DK device.
The specifications bellow only list supported features.

Specifications
--------------

SOC: STM32MP135FAF7
RAM: 512 MiB
Storage: SD Card
Ethernet: 2x 100 Mbps
Wireless: 2.4GHz Cypress CYW43455 (802.11b/g/n)
LEDs: Heartbeat (Blue)
Buttons: 1x Reset, 1x User (USER2)
USB: 4x 2.0 Type-A

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
murata-firmware: add new package

This package contains firmwares provided by Murata.
See https://community.murata.com/.

Murata firmwares are used by wifi chip on stm32 boards.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
murata-nvram: add new package

This package contains nvram files for Murata's Wi-Fi/Bluetooth modules.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke commented:
The CI is failing like this:
```
WARNING: Makefile 'package/utils/audit/Makefile' has a dependency on 'libev', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a build dependency on 'libpam', which does not exist
WARNING: Makefile 'package/boot/kexec-tools/Makefile' has a dependency on 'liblzma', which does not exist
WARNING: Makefile 'package/network/services/lldpd/Makefile' has a dependency on 'libnetsnmp', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a build dependency on 'libpam', which does not exist
make[2]: Entering directory '/__w/openwrt/openwrt/openwrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/__w/openwrt/openwrt/openwrt/scripts/config'
Checking 'python3-cryptography'... ok.
Checking 'python3-pyelftools'... ok.
Checking 'dtc'... time: target/linux/prereq#0.14#0.07#0.19
failed.

optee-os: Please install the device-tree-compiler package

make[3]: *** [/__w/openwrt/openwrt/openwrt/include/prereq.mk:9: prereq] Error 1
Checking 'python3-dev'... ok.
Checking 'python3-setuptools'... ok.
Checking 'swig'... ok.
Package prerequisite check failed.
make[2]: *** [Makefile:86: tmp/.prereq_packages] Error 1
make[1]: *** [/__w/openwrt/openwrt/openwrt/include/toplevel.mk:226: prereq] Error 2
make: *** [/__w/openwrt/openwrt/openwrt/include/toplevel.mk:246: tools/quilt/compile] Error 2
Error: Process completed with exit code 2.
```
Could you have a look at this please.

hauke commented on package/firmware/murata-firmware/Makefile:
```diff
@@ -0,0 +1,57 @@
+PKG_SOURCE_VERSION:=52174a18134c7ef4a674ecd9fb68fc6e2bced969
+
+PKG_MAINTAINER:=Thomas Richard <thomas.richard@bootlin.com>
+PKG_LICENSE_FILES:=LICENCE
```
The license looks ok:
> Subject to the terms and conditions of this
Agreement, Cypress Semiconductor Corporation ("Cypress") grants you a
non-exclusive, non-transferable license under its copyright rights in the
Software to reproduce and distribute the Software in object code form only,
solely for use in connection with Cypress integrated circuit products
("Purpose").

The firmware situation around the cypress chips is a mess. It would be so much easier if this would be in linux-firmware. 
@thom24 This is not your fault and you can probably also not improve this much.

thom24 replied:
I used firmware and configs provided by ST.
It seems murata provides cypress firmware with custom CLM Blob.
There are cypress firmware and CLM Blob in linux-firmware.
But they are older than in cypress repository.

I don't really know how to deal with all of this.
Is it ok for you @hauke to use firmware, CLM blob and config from murata repositories ?
If not what do you expect ?

hauke replied:
yes using the firmware blobs from murata is fine.

hauke commented on package/firmware/murata-firmware/Makefile:
```diff
@@ -0,0 +1,57 @@
+endef
+
+define Package/murata-firmware-43430-sdio/install
+	$(INSTALL_DIR) $(1)/lib/firmware/murata
+	$(INSTALL_DATA) \
+		$(PKG_BUILD_DIR)/cyfmac43430-sdio.bin \
+		$(1)/lib/firmware/murata/
+	$(INSTALL_DATA) \
+		$(PKG_BUILD_DIR)/cyfmac43430-sdio.1DX.clm_blob \
+		$(1)/lib/firmware/murata/
+	$(INSTALL_DIR) $(1)/lib/firmware/brcm
+	$(LN) \
+		../murata/cyfmac43430-sdio.bin \
+		$(1)/lib/firmware/brcm/brcmfmac43430-sdio.bin
+	$(LN) \
+		../murata/cyfmac43430-sdio.1DX.clm_blob \
+		$(1)/lib/firmware/brcm/brcmfmac43430-sdio.clm_blob
+endef
```
What is the difference to the `cypress-firmware-43430-sdio` package?
It also installs `/lib/firmware/brcm/brcmfmac43430-sdio.bin` and will conflict with this package. 

You should define `CONFLICTS:=cypress-firmware-43430-sdio`

thom24 replied:
Murata provides custom CLM Blob.

hauke commented on target/linux/stm32/base-files/lib/preinit/82_brcmfmac_fix_fw_names:
```diff
@@ -0,0 +1,15 @@
+	for ext in bin clm_blob txt
+	do
+		[ ! -f "${fw}.${board}.${ext}" ] && [ -f "${fw}.${ext}" ] && \
+			ln -s "${fw##*/}.${ext}" "${fw}.${board}.${ext}"
```
Why is this needed?
Why don't you directly create the symlinks at build time?

hauke replied:
I think showing the error message is fine if it falls back to the generic file for which you created a symlink.

thom24 replied:
Ok, so I remove it

hauke commented on target/linux/stm32/base-files/lib/preinit/83_switch_to_inband_interrupt_for_wifi_sdio:
```diff
@@ -0,0 +1,9 @@
+
+switch_to_inband_interrupt_for_wifi_sdio() {
+	[ -e /lib/firmware/brcm/brcmfmac43430-sdio.txt ] && \
+		sed -i -E -e 's/^muxenab=0x10/#muxenab=0x10/' "$(readlink -f /lib/firmware/brcm/brcmfmac43430-sdio.txt)"
```
I would prefer to have this at build time.
I think every board should have its unique nvram file.

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,604 @@
+CONFIG_ARM_THUMBEE=y
+CONFIG_ARM_UNWIND=y
+CONFIG_ARM_VIRT_EXT=y
+CONFIG_ASN1=y
```
Why do you need ASN1 support?

thom24 replied:
It was selected by CRYPTO_RSA.
I removed CRYPTO_RSA, so ASN1 is not set anymore.

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,604 @@
+CONFIG_DEBUG_LL_INCLUDE="mach/debug-macro.S"
+CONFIG_DEBUG_MISC=y
+CONFIG_DEBUG_PREEMPT=y
+CONFIG_DEBUG_SECTION_MISMATCH=y
```
Please remove this.

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,604 @@
+CONFIG_FHANDLE=y
+CONFIG_FIXED_PHY=y
+CONFIG_FIX_EARLYCON_MEM=y
+# CONFIG_FORTIFY_SOURCE is not set
```
Do not unset `CONFIG_FORTIFY_SOURCE`

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,604 @@
+CONFIG_GPIO_GENERIC_PLATFORM=y
+CONFIG_GPIO_SYSCON=y
+# CONFIG_GPIO_SYSFS is not set
+# CONFIG_HARDENED_USERCOPY is not set
```
Do not remove `CONFIG_HARDENED_USERCOPY`.

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
The configuration file should be minimal to boot up fine and have network working. All the extra features should be added as modules especially stuff which is optional on boards.
Most options should come from the `target/linux/generic/config-6.6` file and this should only contain which is specifically needed for this SoC.

thom24 replied:
I just did a big cleanup in the config.
If you (or any other people) have other comments on the kernel config, please let me know.

thom24 committed:
kernel: package industrialio-hw-consumer driver

It is needed for the stm32/stm32mp1 subtarget.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: package snd-soc-dmic driver

It is needed for the stm32/stm32mp1 subtarget.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: unset CONFIG_DEVFREQ_THERMAL in kmod-thermal package

It is needed for the stm32/stm32mp1 subtarget.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: package crypto_engine driver

It is needed for the stm32/stm32mp1 subtarget.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: package industrialio-buffer-cb driver

It is needed for the stm32/stm32mp1 subtarget.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: package media controller driver

It is needed for the stm32/stm32mp1 subtarget.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: package v4l2-async and v4l2-fwnode drivers

These packages are needed for the stm32/stm32mp1 subtarget.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: add kmod-media-controller dependency for kmod-usb-audio

The snd-usb-audio drivers uses media controller if media controller is set.
So add kmod-media-controller as dependency.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16716
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 commented:
@hauke : all github checks are now ok

hauke commented on package/utils/dtc/Makefile:
```diff
@@ -102,6 +103,11 @@ define Build/InstallDev
+	$(INSTALL_DIR) $(STAGING_DIR_HOSTPKG)/bin/
+	$(CP) $(MESON_HOST_BUILD_DIR)/dtc $(STAGING_DIR_HOST)/bin/
+endef
+$(eval $(call HostBuild))
```
for TF-A we just use the dtc binary from the linux kernel, see https://github.com/openwrt/openwrt/blob/main/include/trusted-firmware-a.mk#L77.

Where do you need this dtc binary?

thom24 replied:
I needed dtc binary for optee-os package.
And I forgot to include kernel.mk in arm-trusted-firmware-stm32.

Yes I can use dtc binary from linux kernel for optee and I added missing kernel.mk in arm-trusted-firmware-stm32 makefile, so this is not needed anymore.

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_USE_OF=y
+CONFIG_VFP=y
+CONFIG_VFPv3=y
+# CONFIG_VHOST_MENU is not set
+# CONFIG_VIRTIO_MENU is not set
```
Please remove these

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_SWP_EMULATE=y
+CONFIG_SYNC_FILE=y
+# CONFIG_SYN_COOKIES is not set
+CONFIG_SYSFS_SYSCALL=y
```
Please remove this

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_SWPHY=y
+CONFIG_SWP_EMULATE=y
+CONFIG_SYNC_FILE=y
+# CONFIG_SYN_COOKIES is not set
```
Please activate `CONFIG_SYN_COOKIES`.

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_SRAM=y
+CONFIG_SRAM_EXEC=y
+CONFIG_STACKTRACE=y
+# CONFIG_STAGING is not set
```
Please remove this

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_SERIO_SERPORT=y
+CONFIG_SGL_ALLOC=y
+CONFIG_SG_POOL=y
+# CONFIG_SLAB_FREELIST_HARDENED is not set
+# CONFIG_SLAB_FREELIST_RANDOM is not set
```
Please remove these lines

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+# CONFIG_SCHED_STACK_END_CHECK is not set
+CONFIG_SCSI=y
+CONFIG_SCSI_COMMON=y
+# CONFIG_SECTION_MISMATCH_WARN_ONLY is not set
+# CONFIG_SECURITY_DMESG_RESTRICT is not set
```
Please remove these

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_RTC_MC146818_LIB=y
+CONFIG_RWSEM_SPIN_ON_OWNER=y
+CONFIG_SCHED_MC=y
+# CONFIG_SCHED_STACK_END_CHECK is not set
```
Please remove this

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_RATIONAL=y
+# CONFIG_RAVE_SP_CORE is not set
+CONFIG_RCU_CPU_STALL_TIMEOUT=21
+CONFIG_RCU_TRACE=y
```
Please remove this

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_OF_MDIO=y
+CONFIG_OF_OVERLAY=y
+CONFIG_OF_RESOLVE=y
+CONFIG_OID_REGISTRY=y
+CONFIG_OLD_SIGACTION=y
+CONFIG_OLD_SIGSUSPEND3=y
```
Please remove these

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_IRQ_WORK=y
+# CONFIG_ISDN is not set
+CONFIG_JBD2=y
+CONFIG_KALLSYMS=y
```
Please remove this

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,506 @@
+CONFIG_IRQ_DOMAIN_HIERARCHY=y
+CONFIG_IRQ_FORCED_THREADING=y
+CONFIG_IRQ_WORK=y
+# CONFIG_ISDN is not set
```
Please remove this

hauke commented on target/linux/stm32/patches-6.6/900-ARM-dts-stm32-add-ethernet2-for-STM32MP135F-DK-board.patch:
```diff
@@ -0,0 +1,51 @@
+Date: Thu, 26 Sep 2024 16:48:52 +0200
+Subject: [PATCH] ARM: dts: stm32: add ethernet2 for STM32MP135F-DK board
+
+Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
```
Are you planning to add this to upstream Linux kernel too?

hauke replied:
Which of these patches are not yet in linus mainline tree?

thom24 replied:
- 020-mmc-mmci-stm32-add-SDIO-in-band-interrupt-mode.patch

>  Is available upstream, but I cherry-picked the commit from the ST tree. For next iteration, I updated the patch to use the upstream commit.

- 100-ARM-dts-stm32-rtc-add-pin-to-provide-LSCO-on-stm32mp.patch
- 101-ARM-dts-stm32-rtc-add-LSCO-to-WLAN-BT-module-on-stm3.patch
- 102-ARM-dts-stm32-add-support-of-WLAN-BT-on-stm32mp135f-.patch

> These three patches were sent on mailing list, but not accepted, when I created the pull request. Now they are in the pull request sent by Alexandre Torgue for 6.13 (https://lore.kernel.org/all/92d2d6df-cc5c-488f-8ebd-550b1903db12@foss.st.com/). We can assume that they will be merged upstream during next merge window. For next iteration I updated patches to use commits from the pull request and I renamed them as 03x- (we can consider they are upstream backports).

- 700-net-stmmac-dwmac-stm32-add-support-of-phy-supply-pro.patch

> This patch is needed to enable the support of ethernet2. There is nothing available upstream.
I wrote it using ST linux 6.1 as reference.
Now ST has released 6.6 kernel, so there is a patch that I can easily backport.
For next iteration, I updated the patch to use this new one.

- 900-ARM-dts-stm32-add-ethernet2-for-STM32MP135F-DK-board.patch

>Same than for 700-net-stmmac-dwmac-stm32-add-support-of-phy-supply-pro.patch.

For these 2 patches (700 and 900), there is nothing available upstream. But ST planned to do the job (as mentioned in the commit message https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/arch/arm/boot/dts/st/stm32mp135f-dk.dts?id=e9442f1fa4d2545dd6c0aaf7cb7a125cb04f8f2f). So I'm not planning to upstream these two patches.

Other patches are already available upstream.

hauke replied:
Thanks for the update. This looks fine. If they just backports we do not have to maintain the patches for ever.

hauke commented on target/linux/stm32/image/Makefile:
```diff
@@ -0,0 +1,64 @@
+  DEVICE_PACKAGES := kmod-brcmfmac \
+		     murata-firmware-43430-sdio \
+		     murata-nvram-43430-sdio \
+		     kmod-mac80211 \
```
`kmod-mac80211` will be added indirectly not needed here.

thom24 replied:
Good question :)
Indeed this package is not needed

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,485 @@
+CONFIG_SOC_BUS=y
+CONFIG_SOFTIRQ_ON_OWN_STACK=y
+CONFIG_SPARSE_IRQ=y
+CONFIG_SQUASHFS_DECOMP_MULTI_PERCPU=y
+# CONFIG_SQUASHFS_EMBEDDED is not set
+CONFIG_SQUASHFS_FILE_CACHE=y
+# CONFIG_SQUASHFS_FILE_DIRECT is not set
+CONFIG_SQUASHFS_LZO=y
+CONFIG_SQUASHFS_ZLIB=y
```
Do you use squashfs? If it is not used you can also deactivate it, otherwise it is probably better to keep the default settings.

hauke commented on target/linux/stm32/stm32mp1/config-6.6:
```diff
@@ -0,0 +1,485 @@
+CONFIG_OF_MDIO=y
+CONFIG_OF_OVERLAY=y
+CONFIG_OF_RESOLVE=y
+CONFIG_OLD_SIGACTION=y
+CONFIG_OLD_SIGSUSPEND3=y
```
Please remove support for these old signals.

hauke commented:
The build is currently failing in our build bots because it can not find python cryptography:
`Checking 'python3-cryptography'... failed.`
https://buildbot.openwrt.org/images/#/builders/260/builds/1
